src/dpi.rs

```diff
@@ -1,80 +1,85 @@
+//! event whenever a window's scale factor has changed. This can happen if the user drags their
+//! window from a standard-resolution monitor to a high-DPI monitor, or if the user changes their
+//! DPI settings. This gives you a chance to rescale your application's UI elements and adjust how
+//! the platform changes the window's size to reflect the new scale factor. If a window hasn't
+//! received a [`HiDpiFactorChanged`](../enum.WindowEvent.html#variant.HiDpiFactorChanged) event,
+//! then its scale factor is `1.0`.
 //!
-//! The window's logical size is conserved across DPI changes, resulting in the physical size changing instead. This
-//! may be surprising on X11, but is quite standard elsewhere. Physical size changes always produce a
-//! [`Resized`](../event/enum.WindowEvent.html#variant.Resized) event, even on platforms where no resize actually occurs,
-//! such as macOS and Wayland. As a result, it's not necessary to separately handle
-//! [`HiDpiFactorChanged`](../event/enum.WindowEvent.html#variant.HiDpiFactorChanged) if you're only listening for size.
+//! ## How is the scale factor calculated?
 //!
-//! Your GPU has no awareness of the concept of logical pixels, and unless you like wasting pixel density, your
-//! framebuffer's size should be in physical pixels.
+//! Scale factor is calculated differently on different platforms:
 //!
-//! `winit` will send [`Resized`](../enum.WindowEvent.html#variant.Resized) events whenever a window's logical size
-//! changes, and [`HiDpiFactorChanged`](../enum.WindowEvent.html#variant.HiDpiFactorChanged) events
-//! whenever the DPI factor changes. Receiving either of these events means that the physical size of your window has
-//! changed, and you should recompute it using the latest values you received for each. If the logical size and the
-//! DPI factor change simultaneously, `winit` will send both events together; thus, it's recommended to buffer
-//! these events and process them at the end of the queue.
+//! - **Windows:** On Windows 8 and 10, per-monitor scaling is readily configured by users from the
+//!   display settings. While users are free to select any option they want, they're only given a
+//!   selection of "nice" scale factors, i.e. 1.0, 1.25, 1.5... on Windows 7, the scale factor is
+//!   global and changing it requires logging out. See [this article][windows_1] for technical
+//!   details.
+//! - **macOS:** "retina displays" have a scale factor of 2.0. Otherwise, the scale factor is 1.0.
+//!   Intermediate scale factors are never used. It's possible for any display to use that 2.0 scale
+//!   factor, given the use of the command line.
+//! - **X11:** On X11, we calcuate the scale factor based on the millimeter dimensions provided by
+//!   XRandR. This is almost certainly the wrong approach, but many man-hours have been spent trying
+//!   to figure out how to do it better and X11 provides no easy answer.
```

Comment:
We're also using `Xft.dpi` thing and an `env` variable, so I guess you should mention at least `Xft.dpi`.

Reply:
Oh, missed that when going through the source code! Thanks for bringing it up. Honestly, we should probably remove the screen size DPI code and just fall back to `1.0`, since literally nobody expects the current behavior.

Reply:
@chrisduerr will be very unhappy if you remove it and a lot of other alacritty users.

Reply:
besides this joke part. There's no other way around to use multi monitor setups on X11 with different monitors DPI. So with `Xft.dpi` each monitor will have a scaling factor of `Xft.dpi / 96.0`. This is why we brought https://github.com/rust-windowing/winit/issues/836 , it's a great feature to have a actually working multi monitor mixed DPI setup working while having Xft.dpi or something in Xresources, since some apps relies on it.

Reply:
:/

Reply:
Eh, I guess we should just accept that DPI scaling on X11 has always, and will always be, a clusterfuck. I'm just glad I don't have to personally deal with any of it.

Reply:
I think with @kchibisov's PR things work perfectly fine. You have the choice to use Xft.dpi if you need it for other applications, you have the choice to still use dynamic scaling only for winit applications by passing `0` to the env variable, you have the choice to pick any arbitrary factor and if you don't like any of it, you can just fall back to the 'default' X behavior by passing `0`.